Sema: Fix source compatibility break with default initialization of optional properties

slavapestov opened the pull request:
Fixes rdar://problem/35319847.

slavapestov committed:
Sema: Fix source compatibility break with default initialization of optional properties

In Swift 4, properties declared with a sugared Optional type,
like Int?, have a default value of nil. This can be observed
in two ways:

- Classes and structs get an implicit no-argument initializer
- Designated initializers don't have to initialize this property

Note that this did not apply in general to properties where
the type was spelled explicitly as Optional<Int>, etc, mostly
because of implementation restrictions -- when we check if a
type has implicit initializers, we have not realized types for
all stored property members yet, and doing so is not really
possible without the iterative decl checker.

However, in some cases, we *did* perform default initialization
for Optional<Int>, because of some code duplication and
divergent code paths.

A recent refactoring cleaned up some of the mess in this area,
but accidentally broke source compatibility with code that
relied on the broken Optional<Int> case.

Fix this by simulating the old behavior in -swift-version 4,
and preserving the more correct behavior in -swift-version 5.

Fixes <rdar://problem/35319847>.

slavapestov committed:
Sema: Do less work for structs in addImplicitConstructors()

slavapestov committed:
Sema: Fix incorrect usage of NormalProtocolConformance in Codable synthesis

Noticed by inspection. It looks like there are no tests for generic
Codable types. I'm just adding one to cover the changed code path.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

lattner commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -4127,8 +4127,19 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
-          if (!PBD->isDefaultInitializable(i))
-            continue;
+          //
+          // Note: in Swift 4, we screwed up: a property of Optional<Int> does
```
Please don't add comments like this that tell a personal story of the team.  Instead keep it factual:

// Note: The Swift 4 compiler contained a bug that ...

slavapestov committed:
Sema: Type aliases with an unbound generic type cannot witness associated type requirements

We allow definitions like this:

struct G<T> {}
typealias A = G

As a shorthand for

typealias A<T> = G<T>

A typealias like this cannot satisfy an associated type requirement
for the same reason that a generic typealias cannot satisfy an
associated type requirement, which was already handled.

Previously this would crash in the type checker or in IRGen.

This fixes a weird regression in a fixed compiler crasher from the
next patch.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test
